gpflow/conditionals.py

```diff
@@ -124,6 +125,68 @@ def base_conditional(Kmn, Kmm, Knn, f, full_cov=False, q_sqrt=None, whiten=False
     return fmean, fvar
 
 
+@name_scope()
+def uncertain_conditional(Xnew_mu, Xnew_var, feat, kern, q_mu, q_sqrt, whiten=False):
+    """
+    Calculates the conditional for uncertain inputs Xnew, p(Xnew) = N(Xnew_mu, Xnew_var).
+    See ``conditional`` documentation for further reference.
+
+    :param Xnew_mu: mean of the inputs, size N x D
+    :param Xnew_var: covariance matrix of the inputs, size N x D x D
+    :param feat: gpflow.InducingFeature object, only InducingPoints is supported
+    :param kern: gpflow kernel or ekernel object.
+    :param q_mu: mean inducing points, size M x D
+    :param q_sqrt: covariance inducing points, size M x M x D
+    :param whiten: boolean whether to whiten the representation. Default is False.
+
+    :return fmean, fvar: mean and covariance of the conditional, size N x D and N x D x D
+    """
+
+    # TODO: Tensorflow 1.3 doesn't support broadcasting in``tf.matmul`` and
+    # ``tf.matrix_triangular_solve``. This is reported in issue 216.
+    # As a temporary workaround, we are using ``tf.einsum`` for the matrix
+    # multiplications and tiling in the triangular solves.
+    # The code that should be used once the bug is resolved is added in comments.
+
+    assert isinstance(feat, InducingPoints)
+
+    num_data = tf.shape(Xnew_mu)[0] # number of new inputs (N)
+    num_func = tf.shape(q_mu)[1] # output dimension (D)
+
+    q_sqrt_r = tf.matrix_band_part(tf.transpose(q_sqrt, (2, 0, 1)), -1, 0) # D x M x M
+
+    eKuf = tf.transpose(feat.eKfu(kern, Xnew_mu, Xnew_var))  # M x N
+    Kuu = feat.Kuu(kern, jitter=settings.numerics.jitter_level) # M x M
+    Luu = tf.cholesky(Kuu) # M x M
+
+    if not whiten:
+        q_mu = tf.matrix_triangular_solve(Luu, q_mu, lower=True)
+        Luu_tiled = tf.tile(Luu[None, :, :], [num_func, 1, 1]) # remove line once issue 216 is fixed
+        q_sqrt_r = tf.matrix_triangular_solve(Luu_tiled, q_sqrt_r, lower=True)
+
+    Li_eKuf = tf.matrix_triangular_solve(Luu, eKuf, lower=True) # M x N
+    fmean = tf.matmul(Li_eKuf, q_mu, transpose_a=True)
+
+    eKff = kern.eKdiag(Xnew_mu, Xnew_var) # N
+    eKuffu = feat.eKufKfu(kern, Xnew_mu, Xnew_var)  # N x M x M
+    Luu_tiled = tf.tile(Luu[None, :, :], [num_data, 1, 1])  # remove this line, once issue 216 is fixed
+    Li_eKuffu_Lit = tf.matrix_triangular_solve(Luu_tiled, tf.matrix_transpose(eKuffu), lower=True)
+    Li_eKuffu_Lit = tf.matrix_triangular_solve(Luu_tiled, tf.matrix_transpose(Li_eKuffu_Lit), lower=True) # N x M x M
+
+    cov = tf.matmul(q_sqrt_r, q_sqrt_r, transpose_b=True) # D x M x M
+
+    fvar = (
+            tf.matrix_diag(tf.tile((eKff - tf.trace(Li_eKuffu_Lit))[:, None], [1, num_func])) +
+            tf.matrix_diag(tf.trace(tf.einsum("nmj,djk->ndmk", Li_eKuffu_Lit, cov))) +
```

Comment:
is this equivalent to something like tf.einsum("nmj,djm->nd", Li_eKuffu_Lit, cov))?